### PR TITLE
Add common Erlang dirs to search path

### PR DESCRIPTION
**Pull request by lukebakken**
Related to https://github.com/elixir-editors/vim-elixir/pull/543

### CODE REVIEW

**Comment by hcs42** [comment]
Hi,

What is the use case? For example `gf` for `include_lib("filename.erl")`?

Which build system do you use? I use rebar3 these days, and it has different directories (e.g., is has `_build` instead of `deps`, and the umbrella projects also have an `apps` directory).

I think that if we include this, we should have an option for it, e.g. `g:erlang_extend_path`. We can set it to `1` by default if we think that it's useful for most people. See the other options here: https://github.com/vim-erlang/vim-erlang-compiler/blob/b334e956026f61c0bf289ffdf37ce9b2aefe01e1/doc/vim-erlang-compiler.txt#L57

**Comment by lukebakken** [comment]
> What is the use case? For example gf for include_lib("filename.erl")?

I frequently use the `find`, `sf` and `vert sf` commands to open files that I know are in either `src/` or `deps/`.

> Which build system do you use?

That's a good point. I'm a member of the RabbitMQ core engineering team. RabbitMQ uses `erlang.mk`. I also work with `rebar` based projects and it would make sense to update the search directories with `rebar`-specific ones. I'll do that.

I also like the idea of an option, I'll get that as well.

Thanks for the review!

**Comment by lukebakken** [comment]
@hcs42 I've added the rebar directories and configuration setting. It would be great to make the path addition smart enough so that for a project named `foobar` the path `_build/default/lib/foobar` *isn't* added to the search path. I'll see if I can figure out the vimscript for that.

I also need to add tests. Let me know how it looks so far if you have a second!

**Comment by hcs42** [comment]
It looks good so far. (I haven't tested it yet though.)

A few comments:

1.  "It would be great to make the path addition smart enough so that for a project named `foobar` the path `_build/default/lib/foobar` _isn't_ added to the search path. I'll see if I can figure out the vimscript for that." → For umbrella projects, I think this would need to be applied to all `foobar` applications in the `apps` directory. How about just modifying the order to ensure that `apps` directories come first:

    ```
    let &l:path = join(['apps/*/include',
                \       'apps/*/src',
                \       '_build/default/lib/*/src',
                \       '_build/default/*/include', &l:path], ',')
    ```

    This way `:find` and other commands would work correctly even if `foobar` is present in both `apps` and `_build`.

2.  I try to have maximum 80 character long lines where possible. (I haven't reformatted all existing long lines.)

3.  Feel free work in one commit and amend that when you make a change. I prefer merging most contributions as one commit in the end.

4.  Could you update the documentation too before finalizing the PR? It would be good to mention that this feature depends on setting the current directory (`:help current-directory`) to the project's or application's root folder. This is usually not the case in my development workflow (I run only one GVim instance, and do all of my editing inside that). However, I know that there are users who start Vim from the project directory.

**Comment by lukebakken** [comment]
1. Great suggestion.
2. Yep, I'll re-format before rebasing.
3. Will do. I planned on squashing before the final merge.
4. Absolutely!

**Comment by lukebakken** [comment]
@hcs42 - ready for a re-review, thank you.

File: ftplugin/erlang.vim [hunk 1]
```diff
@@ -30,6 +30,26 @@ setlocal commentstring=%%s
...
+        \ 'deps/*/include',
+        \ 'deps/*/src',
+        \ 'deps/*/test',
+        \ 'deps/*/apps/*/include',
+        \ 'deps/*/apps/*/src',
+        \ &l:path], ',')
+  " typical rebar3 paths
+  let &l:path = join([
+        \ 'apps/*/include',
+        \ 'apps/*/src',
```
Comment (lhoguin) [inline ftplugin/erlang.vim:47]: Those are also typical Erlang.mk paths, it's just RabbitMQ doing things weird. :-) See https://github.com/rabbitmq/rabbitmq-server/pull/3133

The apps/ and deps/ come from rebar2 which both Erlang.mk and rebar3 inherit.

Comment (lukebakken) [reply ftplugin/erlang.vim:47]: `deps` isn't RabbitMQ-specific with regard to `erlang.mk`, correct?

Comment (lhoguin) [reply ftplugin/erlang.vim:47]: OK I said something a bit incorrect and I will clarify.

* deps/ is something from rebar2 that Erlang.mk reuses and is not RabbitMQ specific.
* apps/ is something from rebar2 that both Erlang.mk and rebar3 support.
* The way RabbitMQ uses deps is something specific to RabbitMQ. RabbitMQ puts in deps/ what should be under apps/ (as well as the normal deps that should be under deps/).

So what I am trying to say here is that apps/ is not something specific to rebar3, it is something that rebar2/rebar3/Erlang.mk all support, and this is what RabbitMQ *should* use (but currently doesn't).

Comment (lukebakken) [reply ftplugin/erlang.vim:47]: Thanks for the clarification @lhoguin. I just wanted to ensure I wasn't proposing a RabbitMQ-specific change in this PR.

**Comment by hcs42** [comment]
@lukebakken Thank you for your contribution!

**Pull request merged**